src/lib/expression/abstract_expression.hpp

```diff
@@ -127,21 +127,26 @@ inline std::ostream& operator<<(std::ostream& stream, const AbstractExpression&
 
 // Wrapper around expression->hash(), to enable hash based containers containing std::shared_ptr<AbstractExpression>
 struct ExpressionSharedPtrHash final {
-  size_t operator()(const std::shared_ptr<AbstractExpression>& expression) const { return expression->hash(); }
+  size_t operator()(const std::shared_ptr<const AbstractExpression>& expression) const { return expression->hash(); }
```

Comment:
I'd provide both signatures, otherwise I would vouch for there being no copy when this is called with a non-const Expression. Same for ExpressionSharedPtrEqual

Reply:
Not sure if I can follow. If the compiler wouldn't find suitable `ExpressionSharedPtrHash` for `ExpressionUnorderedMap`, wouldn't it complain? This experiment shows that is does convert it to the matching form:
https://godbolt.org/z/E8-Ss7
Are you concerned about the use_count being increased?

Reply:
> Are you concerned about the use_count being increased

Yep, that was my point. Having both a `const std::shared_ptr<AbstractExpression>& expression` as well as a `const std::shared_ptr<const AbstractExpression>& expression` would unequivocally avoid that, right?

Reply:
done.